### DISCUSSION TITLE
My device works in Z2MQTT but cannot control RGB. Solvable?

### DISCUSSION THREAD

**Discussion by teksrq** [Koenkk/zigbee2mqtt#20987] [Devices]
It's been a Zigbee crash course for me over the last week. I currently have several lights working successfully in HA/Z2MQTT, just missing RGB color control. These lights are designed for Smartthings control, I can change colors in the ST iPhone app via Bluetooth, but in Z2MQTT I can only control state, brightness & color temp. I contacted the manufacturer; figured I'd get their boilerplate response, and wasn't surprised. They say color can only be controlled through Smartthings (sounds like Samsung had a big hand in their closed-source firmware). 

I don't want to believe them, but I also don't want to spend days/weeks/months of my life only to find that to be the case. I looked over the "Add New Devices" documentation and into purchasing a nRF Sniffer (to see if I can figure out the Bluetooth commands sent to control RGB), but my rational/adult self is telling me I'm not in that stage of life where jumping down rabbit-holes comes at no personal cost. 

Can anybody with more knowledge of what I'm facing tell me where I'm at? I've been coding for decades, love hacking devices, but I'm not a wiz...and I'm old....

**Comment by teksrq** [comment]
Okay, the crickets were telling me to ask a better question, so I installed MQTTX Desktop to see whatmMQTT payloads I could send to the lights. After looking through the z2m logs and a hour or so of fails, I was able to come up with a payload that controls the lights. The payloads in the logs also contain a "color" property and "color_temp" (there are more properties, but trying to brief); any color I try to set throws the error "***no converter available for 'color'***" while color_temp allows me to set a value from 200 (cool) to 370 (warm). 

So, my question is still much the same, but I think I can now ask it better. I take it that the error message means z2m doesn't (currently) have the ability to covert/create the command that would change the color. Is the lack of a converter for color something that can/might be solvable, or does that (probably) mean that, barring flashing some future custom firmware, these lights simply won't allow zigbee commands that control colors (locking that functionality to/through SmartThings)?

**Reply by tbinitie** [reply]
I don't know if this is helpful, but I previously had full color control with a Sengled bulb in Z2MQTT but lost it after an update. I haven't dedicated to time to fully troubleshooting it yet, but found some folks with a similar problem: https://community.home-assistant.io/t/no-colour-options-on-sengled-bulbs-after-z2mqtt-update/715842

Summarizing the thread: most likely a Z2MQTT problem that has been fixed and will be pushed in a future update.

**Comment by pLavrenov** [comment]
I have the same problem.

In the log I see that when choosing a color it displays a message
Received MQTT message on 'zigbee2mqtt/0x2c116/set' with data '{"color":{"hex":"#FFA52C"}}'

 I changed the config for the device:
`
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const modernExtend = require('zigbee-herdsman-converters/lib/modernExtend');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const legacy = require('zigbee-herdsman-converters/lib/legacy');
const light = require('zigbee-herdsman-converters/lib/light');
const utils = require('zigbee-herdsman-converters/lib/utils');
const presets = exposes.presets;
const access = exposes.access;


const silvercrestModes = {
    white: 0,
    color: 1,
    effect: 2,
};

const hex2rgb = (hex) => {
    const r = parseInt(hex.slice(1, 3), 16);
    const g = parseInt(hex.slice(3, 5), 16);
    const b = parseInt(hex.slice(5, 7), 16);
    return {r, g, b};
}

const rgbToHsb = (r, g, b) => {
    r /= 255;
    g /= 255;
    b /= 255;
    const v = Math.max(r, g, b),
        n = v - Math.min(r, g, b);
    const h =
        n === 0 ? 0 : n && v === r ? (g - b) / n : v === g ? 2 + (b - r) / n : 4 + (r - g) / n;

    return {
        h: 60 * (h < 0 ? h + 6 : h),
        s: v && (n / v) * 100,
        b: v * 100
    }
};

const make4sizedString = (v) => {
    if (v.length >= 4) {
        return v;
    } else if (v.length === 3) {
        return '0' + v;
    } else if (v.length === 2) {
        return '00' + v;
    } else if (v.length === 1) {
        return '000' + v;
    } else {
        return '0000';
    }
};

const fillInHSB = (h, s, b, state) => {
    // Define default values. Device expects leading zero in string.
    const hsb = {
        h: '0168', // 360
        s: '03e8', // 1000
        b: '03e8', // 1000
    };

    if (h) {
        // The device expects 0-359
        if (h >= 360) {
            h = 359;
        }
        hsb.h = make4sizedString(h.toString(16));
    } else if (state.color && state.color.hue) {
        hsb.h = make4sizedString(state.color.hue.toString(16));
    }

    // Device expects 0-1000, saturation normally is 0-100 so we expect that from the user
    // The device expects a round number, otherwise everything breaks
    if (s) {
        hsb.s = make4sizedString(utils.mapNumberRange(s, 0, 100, 0, 1000).toString(16));
    } else if (state.color && state.color.saturation) {
        hsb.s = make4sizedString(utils.mapNumberRange(state.color.saturation, 0, 100, 0, 1000).toString(16));
    }

    // Scale 0-255 to 0-1000 what the device expects.
    if (b != null) {
        hsb.b = make4sizedString(utils.mapNumberRange(b, 0, 255, 0, 1000).toString(16));
    } else if (state.brightness != null) {
        hsb.b = make4sizedString(utils.mapNumberRange(state.brightness, 0, 255, 0, 1000).toString(16));
    }
    return hsb;
};
`
